### ISSUE TITLE
GL Intermediate Representation

### ISSUE THREAD

**Issue by rossant** [vispy/vispy#510]
As discussed with @almarklein, we think it would be a good idea to make some changes in gloo in order to improve the implementation, fix the various issues we had, and considerably simplify the WebGL backends.

The idea is to design a simple language that we call [OpenGL Intermediate Representation](https://github.com/vispy/vispy/wiki/Spec.-Gloo-IR) (GLIR) for now. Basically, it sits between OpenGL and gloo. It defines a few commands for creating, updating, and deleting gloo objects.

The new gloo will be implemented in two parts:
1. The Pythonic gloo user interface.
2. The Python implementation of this interface, compiling to GLIR.

Every gloo function generates a list of GLIR instructions to execute. Then, these instructions are parsed by the Python GLIR interpreter, which executes the actual OpenGL commands. This allows us to decouple the interface from the implementation.

We get a static representation of a scene, that can be interpreted by any language (Python, JavaScript, C++, or other).

Advantages:
- It becomes easier to change either the interface or the implementation without changing the other.
- We can easily impement lazy execution, just by registering the list of GLIR instructions in a FIFO queue, and poping instructions from the queue when we need to draw the scene.
- It simplifies debugging and profiling: we can easily have logging without doing any hack to gloo.
- Writing the WebGL backends becomes much easier. All we have to do is write a JavaScript interpreter for GLIR, which is quite easy. In effect, when using WebGL, Python generates GLIR instructions, but the Python GLIR engine doesn't interprets them. Rather, they are sent directly to the browser to be executed by JavaScript.
- Users can also easily write custom client-side interactivity by generating the proper GLIR instructions from user actions.

Possible cons:
- Performance. There might be a slight performance cost of doing that, but I very much doubt it will become perceptible nor significant. Performance will only become a problem if we have many gloo objects, in which case it will be slow anyway because of the large number of GL commands.
- We need to refactor gloo significantly, but we have to do it anyway... #464 

Tackling this is high priority, because we need to make progress on that before we can do the WebGL backend (which is also high priority at this point).

**TODO**
- [ ] complete GLIR spec
- [ ] implement Pythonic API that compiles to GLIR
- [ ] write Python code that interprets GLIR to desktop OpenGL
- [ ] write JavaScript code that interprets GLIR to WebGL
- [ ] write JavaScript API mimicking Pythonic gloo API, that compiles to GLIR (low priority)

WDYT?
